# DMA Assignment 1 - Data Preprocessing

Data transformations are useful for preparing a dataset for answering a particular question. Part of this process involves generating features from the dataset you find relevant to the question at hand. For this lab, we will be using a Yelp reviews dataset. Each row in the dataset depicts one review along with the features of the review (the reviewer, the review text, etc.). The goal of this lab is to convert this reviews dataset into a reviewers dataset by creating different features of each reviewer.

The submission for this assignment should be done *individually*, although you are allowed to work in groups of 2.



# Environment Setup
Run this cell to setup your environment.

In [163]:
# Importing libraries
import pandas as pd
import math
import numpy as np
print('Libraries Imported')

#DOWNLOADING DATASET
#!wget http://people.ischool.berkeley.edu/~zp/course_datasets/yelp_reviews.csv
#!unzip yelp_reviews.zip
print('Dataset Downloaded: yelp_reviews.csv')
df=pd.read_csv('yelp_reviews.csv')
print(df.head())

print('Setup Complete')

Libraries Imported
Dataset Downloaded: yelp_reviews.csv
     type             business_id                 user_id  stars  \
0  review  mxrXVZWc6PWk81gvOVNOUw  mv7shusL4Xb6TylVYBv4CA      4   
1  review  mxrXVZWc6PWk81gvOVNOUw  0aN5QPhs-VwK2vusKG0waQ      5   
2  review  kK4AzZ0YWI-U2G-paAL7Fg  0aN5QPhs-VwK2vusKG0waQ      5   
3  review  mxrXVZWc6PWk81gvOVNOUw  1JUwyYab-uJzEx_FRd81Zg      5   
4  review  mxrXVZWc6PWk81gvOVNOUw  2Zd3Xy8hUVmZkNg7RyNjhg      4   

                                                text        date  cool_votes  \
0  Definitely try the duck dish.   I rank it amon...  2011-06-13           0   
1  Big Ass Burger was awesome! Great $5 mojitos. ...  2011-06-25           1   
2             Unbelievable sandwiches! Good service.  2011-06-25           0   
3  Awesome, awesome, awesome! My mom and sister a...  2011-07-18           1   
4  I had the ribs they were great.  The beer sele...  2011-07-19           1   

   useful_votes  funny_votes  
0             0        

**Q1: What was the highest number of reviews for any one business_id?**
- For this task, we will need to group the reviews dataset by business_id. This will aggregate data for each business, which is what we need. This can be done using [groupby](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.groupby.html).
    - ```yelp_businesses = yelp_dataset.groupby('business_id').size() ```
    - The .size() counts the number of instances for each business_id, which gives us the number of reviews as each instance in this dataset is a review.
    - The following command will sort this list, after which you can take note of the highest value: ```sorted_yelp_businesses = yelp_businesses.sort(ascending=False,inplace=False)```
    - This approach allows you to see the data structure being used in the sort. A quicker approach to getting the max would be to use the max function: ```max(yelp_businesses)```
<br>
<br>

In [26]:
# YOUR CODE HERE
df_num_review = pd.DataFrame( {'count': df.groupby('business_id').size() }).reset_index()
df_num_review = df_num_review.sort_values(by='count', ascending=False)
print(df_num_review.head())

'''
df_num_review = df.groupby('business_id').size()
df_max = max(df_num_review)
print(df_max)
'''

                  business_id  count
3798   4bEjOyTaDG24SY5TxsaUNQ   4128
24132  YNQgak-ZLtYJQxlDwN-qIg   2234
9249   CZjcFdvJhksq9dy58NVEzw   2057
38803  tFU2Js_nbIZOrnKfYJYBBg   2056
4962   6LM_Klmp3hOP0JmsMCKRqQ   1975


"\ndf_num_review = df.groupby('business_id').size()\ndf_max = max(df_num_review)\nprint(df_max)\n"

###Answer: 4128

**Q2: What was the average number of reviews for a business_id?**

In [36]:
# YOUR CODE HERE
review_sum = df_num_review['count'].sum()
review_id  = len(df_num_review)
review_avg = review_sum/review_id
print(review_avg)

12.63413902163123


###Answer: 12.6341

**Q3: What is the average number of reviews per reviewer?**

In [61]:
# YOUR CODE HERE
review_id_2  = len(df.groupby('user_id').size())
review_sum_2 = len(df)
review_avg_2 = review_sum_2 / review_id_2
print(review_avg_2)

3.188511934933203


###Answer: 3.1885

**Q4: What is the average number of cool votes per reviewer?**

In [49]:
# YOUR CODE HERE
cool_sum_2 = df['cool_votes'].sum()
cool_avg_2 = cool_sum_2 / review_id_2
print(cool_avg_2)

1.2417282785380945


###Answer: 1.2417

**Q5: What is the average number of funny votes per reviewer?**

In [50]:
# YOUR CODE HERE
fun_sum_2 = df['funny_votes'].sum()
fun_avg_2 = fun_sum_2 / review_id_2
print(fun_avg_2)

1.10126486404605


###Answer: 1.1013

**Q6: What is the average number of useful votes per reviewer?**

In [51]:
# YOUR CODE HERE
useful_sum_2 = df['useful_votes'].sum()
useful_avg_2 = useful_sum_2 / review_id_2
print(useful_avg_2)

2.484476138872867


###Answer: 2.4845

**Q7: What is the average of the log of the number of reviews per reviewer?**

In [81]:
# YOUR CODE HERE
df['cnt'] = 1
df_1 = df.groupby('user_id').agg({'cnt':'sum'})
df_1['log_cnt'] = df_1['cnt'].apply(lambda x: math.log(x))
review_sum_3 = df_1['log_cnt'].sum()
review_avg_3 = review_sum_3 / review_id_2
print(review_avg_3)


0.7681766036710075


###Answer: 0.7681766036710075

**Q8: What is the average of the log of the number of cool votes per reviewer?**

In [122]:
# YOUR CODE HERE
df_2 = df.groupby('user_id').agg({'cool_votes':'sum'})
df_2_a = df_2.loc[ df_2['cool_votes']!=0 ]
df_2_a['log_cool'] = df_2_a['cool_votes'].apply(lambda x: math.log(x))
                
cool_sum_3 = df_2_a['log_cool'].sum()
cool_avg_3 = cool_sum_3 / review_id_2
print(cool_avg_3)

0.20849425451212117


C:\Users\user\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


###Answer: 0.20849425451212117

**Q9: What is the average of the log of the number of funny votes per reviewer?**

In [125]:
# YOUR CODE HERE
df_3 = df.groupby('user_id').agg({'funny_votes':'sum'})
df_3_a = df_3.loc[ df_3['funny_votes']!=0 ]
df_3_a['log_fun'] = df_3_a['funny_votes'].apply(lambda x: math.log(x))

fun_sum_3 = df_3_a['log_fun'].sum()
fun_avg_3 = fun_sum_3 / review_id_2
print(fun_avg_3)

0.1905064114576415


C:\Users\user\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


###Answer: 0.1905064114576415

**Q10: What is the average of the log of the number of useful votes per reviewer?**

In [126]:
# YOUR CODE HERE
df_4 = df.groupby('user_id').agg({'useful_votes':'sum'})
df_4_a = df_4.loc[ df_4['useful_votes']!=0 ]
df_4_a['log_fun'] = df_4_a['useful_votes'].apply(lambda x: math.log(x))

useful_sum_3 = df_4_a['log_fun'].sum()
useful_avg_3 = useful_sum_3 / review_id_2
print(useful_avg_3)

0.48005864610840415


C:\Users\user\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


###Answer: 0.48005864610840415

**Q11: Find the average of the percentage of total cool votes out of total votes for each reviewer.**

In [170]:
# YOUR CODE HERE
df_vote = pd.DataFrame( df.groupby('user_id').agg({'cool_votes':'sum', 'funny_votes':'sum', 'useful_votes':'sum'}) )
df_vote['ttl_votes'] = ( df_vote['cool_votes'] + df_vote['funny_votes'] + df_vote['useful_votes'] )
df_vote['cool_percent'] = df_vote['cool_votes'] / df_vote['ttl_votes']  
print(df_vote['cool_percent'].mean()) 

'''
df_cool.columns = df_cool.columns.droplevel(0)
df_cool = df_cool.rename(columns = {'sum':'cool_sum', 'count':'cool_count'})
df_cool['cool_percent'] = 100 * ( df_cool['cool_sum'] / df_cool['cool_count'] )
'''

0.1927040524186172


"\ndf_cool.columns = df_cool.columns.droplevel(0)\ndf_cool = df_cool.rename(columns = {'sum':'cool_sum', 'count':'cool_count'})\ndf_cool['cool_percent'] = 100 * ( df_cool['cool_sum'] / df_cool['cool_count'] )\n"

###Answer: 0.1927040524186172

**Q12: Find the average of the percentage of total funny votes out of total votes for each reviewer.**

In [171]:
# YOUR CODE HERE
df_vote['fun_percent'] = df_vote['funny_votes'] / df_vote['ttl_votes']  
print(df_vote['fun_percent'].mean()) 

0.18256423845803052


###Answer: 0.18256423845803052

**Q13: Find the average of the percentage of total useful votes out of total votes for each reviewer.**

In [172]:
# YOUR CODE HERE
df_vote['use_percent'] = df_vote['useful_votes'] / df_vote['ttl_votes']  
print(df_vote['use_percent'].mean()) 

0.6247317091233844


###Answer: 0.6247317091233844

**Q14: Average review text length (in non-space characters)**

In [142]:
# YOUR CODE HERE
df['text_s'] = list( df['text'][i].replace(" ", "") for i in range(len(df['text'])))
df['text_len'] = list( len(df['text_s'][i]) for i in range(len(df['text_s'])) ) 
sum_text = df.agg({'text_len':'sum'})
num_instance = len(df['text_len'])
print(sum_text / num_instance)

text_len    499.025326
dtype: float64


###Answer: 499.025326

**Q15: Year in which the reviewer wrote the most reviews. Once you have this for each reviewer, subtract the minimum possible year (2005) from each so that your final feature values are 0, 1, 2 etc.**

In [44]:
# YOUR CODE HERE
df['year'] = pd.DatetimeIndex(df['date']).year - 2005
df_year = pd.get_dummies(df, columns=['year']).groupby(['user_id'], as_index=False).sum()
df_year['max_review_year'] = df_year.loc[:,'year_-1':'year_11'].idxmax(axis=1)
print( df_year.head() )

                  user_id  stars  cool_votes  useful_votes  funny_votes  \
0  --1Y03CEKR3WDbBjYnsW7A      5           0             0            0   
1  --2QZsyXGz1OhiD4-0FQLQ     14           0             1            0   
2  --82_AVgRBsLw6Dhy8sEnA      4           0             0            0   
3  --8A9o_NeGyt_3kzlXtSdg     11           0             1            0   
4  --8WbseBk1NjfPiZWjQ-XQ     11           0             0            0   

   year_-1  year_0  year_1  year_2  year_3  year_4  year_5  year_6  year_7  \
0      0.0     0.0     0.0     0.0     0.0     0.0     0.0     1.0     0.0   
1      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2      0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0     0.0   
3      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
4      0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

   year_8  year_9  year_10  year_11 max_review_year  
0     0.0     0.0      0.0

###Answer:

**Q16: Come up with a new feature. This may be derived from existing features. Give your feature the name ** *my_new_feature.* **Dsiplay head() of this new feature.**

In [59]:
# YOUR CODE HERE
df_star = pd.DataFrame(df.groupby('business_id').agg({'stars':'mean'})) 
df_star = df_star.rename(columns = {'stars':'my_new_feature'})
df_star = df_star.sort_values('my_new_feature', ascending=False)
print( df_star.head() )

                        my_new_feature
business_id                           
zzxb0YNkRvH1UgifuR3BvQ             5.0
Bfp0U0lrD79X9clT8gS7YQ             5.0
XaxsJK1AEalzgAI0jAVY5w             5.0
XaSV2CimMDKOuNrfP8F6CQ             5.0
Bg3qwBTCsLzqRXT2M-7R5g             5.0


###Answer: